nexus/src/app/mod.rs

```diff
@@ -0,0 +1,471 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Nexus, the service that operates much of the control plane in an Oxide fleet
+
+use crate::authn;
+use crate::authz;
+use crate::config;
+use crate::context::OpContext;
+use crate::db;
+use crate::db::lookup::LookupPath;
+use crate::db::model::Name;
+use crate::populate::populate_start;
+use crate::populate::PopulateStatus;
+use crate::saga_interface::SagaContext;
+use crate::sagas;
+use anyhow::anyhow;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use slog::Logger;
+use std::sync::Arc;
+use uuid::Uuid;
+
+// The implementation of Nexus is large, and split into a number of submodules
+// by resource.
+mod disk;
+mod image;
+mod instance;
+mod organization;
+mod oximeter;
+mod project;
+mod rack;
+mod saga;
+mod session;
+mod silo;
+mod sled;
+// TODO: We should strongly consider moving all tests which depend on these
+// interfaces into unit tests, rather than integration tests.
+//
+// This would enable us to conditionally compile the test interfaces out of
```

Comment:
I don't think I agree with this TODO.  As an example, right now, we have integration tests that create an Instance and wait for it to become booted.  We use these private interfaces to poke the simulated sled agent to boot the Instance.  What should we do instead?

Reply:
For clarity, I think having an interface to do such a thing is reasonable. However, this TODO specifically recommends that if we have integration tests which rely on private interfaces, they aren't really acting as integration tests:

https://doc.rust-lang.org/book/ch11-03-test-organization.html
> Integration tests are entirely external to your library and use your code in the same way any other external code would, using only the public interface and potentially exercising multiple modules per test.

By exposing a "test-only" interface through a module that is *not* labelled `#[cfg(test)]`, it's actually just an interface, like any other - we will be shipping this code with Nexus, so even if we attempt to avoid calling it, it will be present in the release binary.

Reply:
I think if we're adhering strictly to the definitions you linked to, then Cargo only supports:

- "unit tests": white-box unit testing (maybe you could do black-box unit testing by defining tests a level up the module hierarchy so that your test has no access to non-pub items?)
- "integration tests": black-box integration testing (for the reason you pointed out)

But [white-box integration tests](https://en.wikipedia.org/wiki/White-box_testing) are a thing.  That's closer to what we're doing here.  I don't think the tests we're talking about make sense as unit tests since they set up a whole database, Nexus (including its HTTP server), etc.  Yet I still think they're useful because they test the behavior end-to-end.

You're right that we have to work around Cargo's tests being black-box-only by including some test-only interfaces in the built binary and that's not ideal.  Alternatives might be:

- invent a new test runner, maybe by extending Cargo to support a third kind of test that's built against the crate as built with `#[cfg(test)]`
- do white-box integration testing by defining a new crate that depends on the crate we want to test using a Cargo feature that turns on the test interfaces

I'm not sure any of this is a great use of effort but maybe I'm not fully understanding the concrete problem.  I get that we have unused code in the release binary.

Maybe file an issue instead of the TODO?  I find TODOs like this quite distracting.

Reply:
Okay, I'll remove it.